### PR TITLE
Add modules online_server_install and online_manage_c14

### PR DESCRIPTION
**Pull request by broferek**
##### SUMMARY

Add a module to be able to launch a server installation in Online Cloud solution

##### ISSUE TYPE

- New Module Pull Request

##### COMPONENT NAME

online_server_install : Install a server with an OS

##### ANSIBLE VERSION

```paste below
ansible 2.8.0.dev0 (devel 11ead1885b) 
```

##### ADDITIONAL INFORMATION

```paste below

```

### CODE REVIEW

**Comment by broferek** [comment]
@sieben Remy, here is the first version of the module.

It works fine for me. As stated, it only handles partitioning template for the moment.

Thanks to give me feedback

File: lib/ansible/module_utils/online.py [hunk 1]
```diff
@@ -70,11 +70,21 @@ def get_resources(self):
         if not results.ok:
             raise OnlineException('Error fetching {0} ({1}) [{2}: {3}]'.format(
                 self.name, '%s/%s' % (self.module.params.get('api_url'), self.name),
-                results.status_code, results.json['message']
+                results.status_code, results.json['error']
             ))
 
         return results.json
 
+    def post_resources(self, data):
```
Comment (remyleone) [inline lib/ansible/module_utils/online.py:78]: Why not use a post method and have this treatment done elsewhere?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+version_added: "2.8"
+author:
+  - "Ludovic Petetin (@broferek)"
+extends_documentation_fragment: online
+'''
+
+EXAMPLES = r'''
+- name: Install Centos on the server 5824 with the user foo created
+  online_server_install:
+    api_token: '0d1627e8-bbf0-44c5-a46f-5c4d3aef033f'
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py]: Could you have a dummy UUID instead?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    host: "5824"
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py:62]: Why is the host a number?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+    os_id: 305
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py:63]: How did you obtained this ID?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+    hostname: "sd-5824"
+    root_password: "root_bar"
+    user_login: "foo"
+    user_password: "bar"
+    partitioning_template_ref: "a645ex65-af43-4rr8-bc45-654qsdf654e2"
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py]: Where does this template reference come from?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+'''
+
+RETURN = r'''
+---
+msg:
+    description: Message specifying that the installation has started
+    returned: success
+    type: string
+    sample: The installation has started. Check your server state at https://api.online.net/fr/server/state/5824
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py:77]: Ansible module are supposed to give a JSON back. Could you add more information id, and other relevant attributes you would get in the payload of the request?

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    host = module.params['host']
+    post_data = dict(        
+        os_id = module.params['os_id'],
+        hostname = module.params['hostname'],
+        root_password = module.params['root_password'],
+        user_login = module.params['user_login'],
+        user_password = module.params['user_password'],
+        partitioning_template_ref = module.params['partitioning_template_ref']
+    )
+    try:
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py:143]: Could you have a shorter try block? I cannot see what function is supposed to raise an exception here.

**Review by remyleone — CHANGES_REQUESTED** [review]

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 2 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: I took the same than you put in [online_server_facts](https://github.com/broferek/ansible/blob/devel/lib/ansible/modules/cloud/online/online_server_facts.py), should I change both hence?

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 3 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:62]: Because the API only servers by their ID, what would you put instead?

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 4 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:63]: I have added documentation about options stating how you retrieve this ID from the API in this [commit](https://github.com/ansible/ansible/pull/46528/commits/3a5ab3e8cd062635942f5d1cb3e231fb8a696124), is that ok?

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 5 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: I have added documentation about options stating how you retrieve this reference this [commit](https://github.com/ansible/ansible/pull/46528/commits/3a5ab3e8cd062635942f5d1cb3e231fb8a696124), is that ok?

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 2 above)
Comment (remyleone) [reply lib/ansible/modules/cloud/online/online_server_install.py]: No only this one. I will change the online_server_facts later I've learnt only recently that you are suppose to have dummy one :)

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 6 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:77]: The API doesn't return anything except a code result, hence, if the call is successful, the ouput message will redirect the user to the console so that he can follow the installation of his server.
Could you be more precise about what you would like to have in the output?

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 7 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:143]: This is actually for both API calls (get_resources and post_resources), I will fix the try catch around these 2 lines only

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 2 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: Done

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 7 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:143]: I have put the try catch around the 2 API calls

File: lib/ansible/module_utils/online.py (hunk 1 above)
Comment (broferek) [reply lib/ansible/module_utils/online.py:78]: I tried to use the same logic than you did with get and get_resources methods, am I wrong?

**Comment by broferek** [comment]
Any chance to merge this pull request?

I have used it again today and it does what the expected job.

Thanks,

Ludo

File: lib/ansible/modules/cloud/online/online_server_install.py [hunk 8]
```diff
@@ -0,0 +1,172 @@
...
+- name: Install Centos on the server 5824 with the user foo created
+  online_server_install:
+    api_token: "qsfda9964-foo-sghz8751-bar"
+    host: "5824"
+    os_id: 305
+    hostname: "sd-5824"
+    root_password: "root_bar"
+    user_login: "foo"
+    user_password: "bar"
+    partitioning_template_ref: "ax686-foo-qsdf84-bar"
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_server_install.py:68]: Could you put a dummy UUID. Because it is not clear whether it is the name or the UUID we should use here.

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 4 above)
Comment (remyleone) [reply lib/ansible/modules/cloud/online/online_server_install.py:63]: Yes it is fine

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 3 above)
Comment (remyleone) [reply lib/ansible/modules/cloud/online/online_server_install.py:62]: I would prefer to have only the hostname used here. You could use online_server_facts to gather it.

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 6 above)
Comment (remyleone) [reply lib/ansible/modules/cloud/online/online_server_install.py:77]: You still need to output a JSON could be something as simple as:

```
{"online_server_install": {"status": "installed"}}
```

Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: I'm still not sure to understand what you mean here.

I use the normal module.exit_json function in case of a module success.
    module.exit_json(
        changed=True,
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )
So, it outputs JSON as you would expect.

The documentation only says that inside the JSON returned, there is a field called "msg" which is a string but the output itself is not a string.

I can add another 'status' field in the exit JSON if you want. Maybe something like this would be better ?
    module.exit_json(
        changed=True,
        status='installing',
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )

I can't say "status installed" as you said as the API returns a 200 code when the installation starts, it doesn't wait or the server to be actually installed. I guess the user will have to wait for the SSH port to be opened in another task to consider its server installed.

Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: I'm still not sure to understand what you mean here.

I use the normal module.exit_json function in case of a module success.
    module.exit_json(
        changed=True,
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )
So, it outputs JSON as you would expect.

The documentation only says that inside the JSON returned, there is a field called "msg" which is a string but the output itself is not a string.

I can add another 'status' field in the exit JSON if you want. Maybe something like this would be better ?
    module.exit_json(
        changed=True,
        status='installing',
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )

I can't say "status installed" as you said as the API returns a 200 code when the installation starts, it doesn't wait or the server to be actually installed. I guess the user will have to wait for the SSH port to be opened in another task to consider its server installed.

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 3 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: Online_server_facts is quite long to gather all hosts facts whereas in my case, I only want one server facts.

If we could filter facts for a specific host then it could be a solution. Without it, I don't think it is a good idea to use it for people with large inventories.

Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py]: Online_server_facts is quite long to gather all hosts facts whereas in my case, I only want one server facts.

If we could filter facts for a specific host then it could be a solution. Without it, I don't think it is a good idea to use it for people with large inventories.

Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:62]: Online_server_facts is quite long to gather all hosts facts whereas in my case, I only want one server facts.

If we could filter facts for a specific host then it could be a solution. Without it, I don't think it is a good idea to use it for people with large inventories.

File: lib/ansible/modules/cloud/online/online_server_install.py (hunk 6 above)
Comment (broferek) [reply lib/ansible/modules/cloud/online/online_server_install.py:77]: I'm still not sure to understand what you mean here.

I use the normal module.exit_json function in case of a module success.
```python
    module.exit_json(
        changed=True,
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )
```
So, it outputs JSON as you would expect.

The documentation only says that inside the JSON returned, there is a field called "msg" which is a string but the output itself is not a string.

I can add another 'status' field in the exit JSON if you want. Maybe something like this would be better ?
```python
    module.exit_json(
        changed=True,
        status='installing',
        msg="The installation has started. Check your server state at {0}/fr/server/state/{1}".format(module.params.get('api_url'), host)
    )
```

I can't say "status installed" as you said as the API returns a 200 code when the installation starts, it doesn't wait or the server to be actually installed. I guess the user will have to wait for the SSH port to be opened in another task to consider its server installed.

**Review by broferek — COMMENTED** [review]
@sieben could you please review my feedbacks?

**Comment by remyleone** [comment]
Could you rebase and squash?

**Comment by broferek** [comment]
@sieben I rebased and squashed. I also deleted duplicated comments in the review. Hope it helps.

Let me know

File: lib/ansible/modules/utilities/logic/import_role.py [hunk 9]
```diff
@@ -48,6 +48,11 @@
       - Overrides the role's metadata setting to allow using a role more than once with the same parameters.
```
Comment (remyleone) [inline lib/ansible/modules/utilities/logic/import_role.py:48]: What is the purpose of this change? Are they required for testing Online module? Could they be reviewed/merged in a different PR?

Comment (broferek) [reply lib/ansible/modules/utilities/logic/import_role.py:48]: Sorry, this is messed up from a rebase, I think.

I think I will close this PR and open a new one when I have worked again on this module. It needs better integration on our side before trying to submit it again.

FYI, we are also working on a C14 module which is pretty good at the moment. I think I will try to submit the C14 first. I'll put you in the process review as well

File: lib/ansible/modules/cloud/online/online_manage_c14.py [hunk 10]
```diff
@@ -0,0 +1,496 @@
...
+    operation = module.params['operation']
+    archive_uuid = module.params['archive_uuid']
+    paths =  module.params['paths']
+    download_path =  module.params['download_path']
+    archive_details =  module.params['archive_details']    
+    archive_key =  module.params['archive_key']
+    wait = module.params['wait']
+    
+    try:
+        if operation == 'get_archives':
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_manage_c14.py]: Instead of having very long `elif` loop. Could you have a dict that got the name of an operation as a key and the function to apply as a value. That way we avoid doing a good deal of useless string comparisons.

File: lib/ansible/modules/cloud/online/online_manage_c14.py [hunk 11]
```diff
@@ -0,0 +1,496 @@
...
+        else:
+            self.module.exit_json(result=result.json)          
+        
+def main():
+    argument_spec = online_argument_spec()
+    argument_spec.update(
+        dict(
+            operation=dict(
+                required=True,
+                choices=['get_archives','get_safes','get_bucket','get_files','get_archive_key','set_archive_key','delete_archive_key','create_archive','upload_files','download_files','freeze_archive','unfreeze_archive','remove_archive','remove_files']
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_manage_c14.py]: Could you put all those strings in a dict as keys and for the values just use a dict. By doing so, you are sure to modify only in one place the name of functions

File: lib/ansible/modules/cloud/online/online_manage_c14.py [hunk 12]
```diff
@@ -0,0 +1,496 @@
...
+        post_data = dict()
+        result = self.post('/%s' % self.name, post_data)
+        time.sleep(5)
+        if wait:
+            while True:
+                archive = self.get_archives(archive_uuid)
+                if archive[0]['status'] == 'active':
+                    break
+                else:
+                    time.sleep(30)
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_manage_c14.py]: Could you have this value configured?

File: lib/ansible/modules/cloud/online/online_manage_c14.py [hunk 13]
```diff
@@ -0,0 +1,496 @@
...
+
+        self.name = 'storage/c14/safe/{0}/archive'.format(archive_details['safe_uuid'])
+            
+        post_data = dict(
+            safe_id=archive_details['safe_uuid'],
+            name=archive_details['name'],
+            description=archive_details['description'],
+            parity=archive_details['parity'] if archive_details.has_key('parity') is None else 'standard',
+            crypto=archive_details['crypto'] if archive_details.has_key('crypto') is None else 'aes-256-cbc',
+            days=archive_details['days'] if archive_details.has_key('days') is None else 7,
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_manage_c14.py]: Could this be configured?

File: lib/ansible/modules/cloud/online/online_manage_c14.py [hunk 14]
```diff
@@ -0,0 +1,496 @@
...
+)
+import pysftp
+import math
+import os
+import time
+import shutil
+import posixpath
+from stat import S_ISDIR
+
+def convert_size(size_bytes):
```
Comment (remyleone) [inline lib/ansible/modules/cloud/online/online_manage_c14.py]: Could you add some documentation to this function? What does it do?

**Review by remyleone — COMMENTED** [review]
Could you put some test for your modules? Like a playbook that do the complete life cycle of an archive and use the different operations to ensure a good coverage of your code?

**Comment by remyleone** [comment]
I think I reviewed at the wrong place. Anyhow you still got my comments to
help you moving your C14 module forward
